### PR TITLE
consul naming support and test case

### PR DESCRIPTION
**Pull request by Moriadry**
see https://github.com/baidu/brpc-java/issues/31 pls help review this, thank you. @wenweihu86

### CODE REVIEW

File: brpc-java-naming-consul/pom.xml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.baidu</groupId>
+        <artifactId>brpc-java-parent</artifactId>
+        <version>2.2.1</version>
```
Comment (wenweihu86) [inline brpc-java-naming-consul/pom.xml]: please use master pom version: 2.2.4-SNAPSHOT

**Comment by guohao** [comment]
Add the license header comment to all new `.java` files

**Comment by wenweihu86** [comment]
please check the compatibility with brpc c++ consul converson. see the detail:
https://github.com/apache/incubator-brpc/blob/master/src/brpc/policy/consul_naming_service.cpp

**Comment by Moriadry** [comment]
License has already been declared. And I will try to figure the compatibility out. @wenweihu86

File: brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulConstants.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+public class ConsulConstants {
+
+    public static final int DEFAULT_CONSUL_INTERVAL = 30;
+
+    public static final String CONSULINTERVAL = "consulInterval";
+
+    public static final String LOOKUPINTERVAL = "lookupInterval";
+
+    /**
+     * service 最长存活周期（Time To Live），单位秒。 每个service会注册一个ttl类型的check，在最长TTL秒不发送心跳 就会将service变为不可用状态。
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulConstants.java]: Let's comment this in English.

File: brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulResponse.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    public Long getConsulLastContact() {
+        return consulLastContact;
+    }
+
+    public void setConsulLastContact(Long consulLastContact) {
+        this.consulLastContact = consulLastContact;
+    }
+
+
+}
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulResponse.java]: Add a new line at the end of this file.

File: brpc-java-naming-consul/src/test/resources/log4j2.xml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    </Appenders>
+    <Loggers>
+        <Logger name="com.baidu.brpc" level="debug" additivity="false">
+            <AppenderRef ref="Console"/>
+        </Logger>
+        <Root level="info">
+            <AppenderRef ref="Console"/>
+        </Root>
+    </Loggers>
+</Configuration>
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/test/resources/log4j2.xml]: Also here.

File: brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/ConsulNamingService.java [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+import com.ecwid.consul.v1.health.model.HealthService;
+import io.netty.util.HashedWheelTimer;
+import io.netty.util.Timeout;
+import io.netty.util.Timer;
+import io.netty.util.TimerTask;
+import io.netty.util.internal.ConcurrentSet;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections4.CollectionUtils;
+
+import java.util.*;
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/ConsulNamingService.java]: Don't use `*`.

File: brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/ConsulNamingService.java [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+import java.util.concurrent.*;
+
+@Slf4j
+public class ConsulNamingService implements NamingService {
+
+    private BrpcURL      url;
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/ConsulNamingService.java]: Let's keep one space as indent.

File: brpc-java-naming-consul/src/test/resources/log4j2.xml (hunk 4 above)
Comment (Moriadry) [reply brpc-java-naming-consul/src/test/resources/log4j2.xml]: thank you for your review, will fix it ASAP :)

**Comment by jovany-wang** [comment]
Could u plz resolve the conflicts?

**Comment by Moriadry** [comment]
@jovany-wang  done.

**Comment by wenweihu86** [comment]
I think it can be merged into master.
@jovany-wang what's your opinion?

File: brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/ConsulNamingService.java (hunk 5 above)
Comment (jovany-wang) [reply brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/ConsulNamingService.java]: This comment have not addressed yet.

File: brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulConstants.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    public static final int DEFAULT_CONSUL_INTERVAL = 30;
+
+    public static final String CONSULINTERVAL = "consulInterval";
+
+    public static final String LOOKUPINTERVAL = "lookupInterval";
+
+    /**
+     * Time To Live, in seconds. Every service will register a ttl for health check.
+     */
+    public static final int TTL                       = 30;
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulConstants.java]: just remaining 1 space before `=`

File: brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulResponse.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.baidu.brpc.naming.consul.model;
+
+public class ConsulResponse<T> {
+
+    /**
+     * consul返回的具体结果
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/main/java/com/baidu/brpc/naming/consul/model/ConsulResponse.java]: In English?

File: brpc-java-naming-consul/src/test/java/com/baidu/brpc/naming/consul/ConsulNamingServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+@Slf4j
+public class ConsulNamingServiceTest {
+
+    private static BrpcURL             namingUrl;
+    private static ConsulNamingService consulNamingService;
+    private static ConsulProcess       consul;
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        String customConfiguration =
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/test/java/com/baidu/brpc/naming/consul/ConsulNamingServiceTest.java]: final for local or `static final` for member.

File: brpc-java-naming-consul/src/test/java/com/baidu/brpc/naming/consul/ConsulNamingServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+    protected SubscribeInfo createSubscribeInfo(boolean ignoreFail) {
+        SubscribeInfo subscribeInfo = new SubscribeInfo();
+        subscribeInfo.setService(EchoService.class.getName());
+        subscribeInfo.setIgnoreFailOfNamingService(ignoreFail);
+        return subscribeInfo;
+    }
+
+    @Test
+    public void testRegisterAndSubscribe() throws Exception {
```
Comment (jovany-wang) [inline brpc-java-naming-consul/src/test/java/com/baidu/brpc/naming/consul/ConsulNamingServiceTest.java]: Don't throws such abstract `Exception`.

**Comment by jovany-wang** [comment]
Very thanks for your contributions!

**Comment by Moriadry** [comment]
My pleasure :)

**Pull request merged**